--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -43,6 +43,15 @@
    private int field_92055_b;
    private LivingEntity field_191513_e;
 
+   // Spigot Start - copied from tick
+   @Override
+   public void inactiveTick() {
+      this.field_92056_a += 1;
+
+      super.inactiveTick();
+   }
+   // Spigot End
+
    public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> p_i50164_1_, World p_i50164_2_) {
       super(p_i50164_1_, p_i50164_2_);
    }
@@ -156,6 +165,13 @@
 
    }
 
+   @Override
+   protected void func_70227_a(RayTraceResult result) {
+      if (result.func_216346_c() == RayTraceResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.func_70227_a(result);
+      }
+   }
+
    private void func_213893_k() {
       this.field_70170_p.func_72960_a(this, (byte)17);
       this.func_191510_k();
