--- a/net/minecraft/world/server/ServerWorld.java
+++ b/net/minecraft/world/server/ServerWorld.java
@@ -143,10 +143,11 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class ServerWorld extends World implements ISeedReader {
+public class ServerWorld extends World implements ISeedReader, net.minecraftforge.common.extensions.IForgeWorldServer {
    public static final BlockPos field_241108_a_ = new BlockPos(100, 50, 0);
    private static final Logger field_147491_a = LogManager.getLogger();
    private final Int2ObjectMap<Entity> field_217498_x = new Int2ObjectLinkedOpenHashMap<>();
+   final com.tuinity.tuinity.util.maplist.IteratorSafeOrderedReferenceSet<Entity> entitiesForIteration = new com.tuinity.tuinity.util.maplist.IteratorSafeOrderedReferenceSet<>(2048, 0.5f, 2048, 0.2, true); // Tuinity - make removing entities while ticking safe
    private final Map<UUID, Entity> field_175741_N = Maps.newHashMap();
    private final Queue<Entity> field_217499_z = Queues.newArrayDeque();
    private final List<ServerPlayerEntity> field_217491_A = Lists.newArrayList();
@@ -173,6 +174,7 @@
    private final DragonFightManager field_241105_O_;
    private final StructureManager field_241106_P_;
    private final boolean field_241107_Q_;
+   private net.minecraftforge.common.util.WorldCapabilityData capabilityData;
 
    public ServerWorld(MinecraftServer p_i241885_1_, Executor p_i241885_2_, SaveFormat.LevelSave p_i241885_3_, IServerWorldInfo p_i241885_4_, RegistryKey<World> p_i241885_5_, DimensionType p_i241885_6_, IChunkStatusListener p_i241885_7_, ChunkGenerator p_i241885_8_, boolean p_i241885_9_, long p_i241885_10_, List<ISpecialSpawner> p_i241885_12_, boolean p_i241885_13_) {
       super(p_i241885_4_, p_i241885_5_, p_i241885_6_, p_i241885_1_::func_213185_aS, false, p_i241885_9_, p_i241885_10_);
@@ -200,7 +202,7 @@
       } else {
          this.field_241105_O_ = null;
       }
-
+      this.initCapabilities();
    }
 
    public void func_241113_a_(int p_241113_1_, int p_241113_2_, boolean p_241113_3_, boolean p_241113_4_) {
@@ -272,17 +274,17 @@
 
          this.field_73018_p = this.field_73017_q;
          if (this.field_72986_A.func_76061_m()) {
-            this.field_73017_q = (float)((double)this.field_73017_q + 0.01D);
+            this.field_73017_q = (float) ((double) this.field_73017_q + 0.01D);
          } else {
-            this.field_73017_q = (float)((double)this.field_73017_q - 0.01D);
+            this.field_73017_q = (float) ((double) this.field_73017_q - 0.01D);
          }
 
          this.field_73017_q = MathHelper.func_76131_a(this.field_73017_q, 0.0F, 1.0F);
          this.field_73003_n = this.field_73004_o;
          if (this.field_72986_A.func_76059_o()) {
-            this.field_73004_o = (float)((double)this.field_73004_o + 0.01D);
+            this.field_73004_o = (float) ((double) this.field_73004_o + 0.01D);
          } else {
-            this.field_73004_o = (float)((double)this.field_73004_o - 0.01D);
+            this.field_73004_o = (float) ((double) this.field_73004_o - 0.01D);
          }
 
          this.field_73004_o = MathHelper.func_76131_a(this.field_73004_o, 0.0F, 1.0F);
@@ -296,15 +298,19 @@
          this.field_73061_a.func_184103_al().func_232642_a_(new SChangeGameStatePacket(SChangeGameStatePacket.field_241772_i_, this.field_73017_q), this.func_234923_W_());
       }
 
+      /* The function in use here has been replaced in order to only send the weather info to players in the correct dimension,
+       * rather than to all players on the server. This is what causes the client-side rain, as the
+       * client believes that it has started raining locally, rather than in another dimension.
+       */
       if (flag != this.func_72896_J()) {
          if (flag) {
-            this.field_73061_a.func_184103_al().func_148540_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241766_c_, 0.0F));
+            this.field_73061_a.func_184103_al().func_232642_a_(new SChangeGameStatePacket(SChangeGameStatePacket.field_241766_c_, 0.0F), this.func_234923_W_());
          } else {
-            this.field_73061_a.func_184103_al().func_148540_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241765_b_, 0.0F));
+            this.field_73061_a.func_184103_al().func_232642_a_(new SChangeGameStatePacket(SChangeGameStatePacket.field_241765_b_, 0.0F), this.func_234923_W_());
          }
 
-         this.field_73061_a.func_184103_al().func_148540_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241771_h_, this.field_73004_o));
-         this.field_73061_a.func_184103_al().func_148540_a(new SChangeGameStatePacket(SChangeGameStatePacket.field_241772_i_, this.field_73017_q));
+         this.field_73061_a.func_184103_al().func_232642_a_(new SChangeGameStatePacket(SChangeGameStatePacket.field_241771_h_, this.field_73004_o), this.func_234923_W_());
+         this.field_73061_a.func_184103_al().func_232642_a_(new SChangeGameStatePacket(SChangeGameStatePacket.field_241772_i_, this.field_73017_q), this.func_234923_W_());
       }
 
       if (this.field_73068_P && this.field_217491_A.stream().noneMatch((p_241132_0_) -> {
@@ -312,8 +318,8 @@
       })) {
          this.field_73068_P = false;
          if (this.func_82736_K().func_223586_b(GameRules.field_223607_j)) {
-            long l = this.field_72986_A.func_76073_f() + 24000L;
-            this.func_241114_a_(l - l % 24000L);
+            long l = this.func_72820_D() + 24000L;
+            this.func_241114_a_(net.minecraftforge.event.ForgeEventFactory.onSleepFinished(this, l - l % 24000L, this.func_72820_D()));
          }
 
          this.func_229856_ab_();
@@ -349,66 +355,57 @@
          }
 
          this.field_217492_a = true;
-         ObjectIterator<Entry<Entity>> objectiterator = this.field_217498_x.int2ObjectEntrySet().iterator();
+         //ObjectIterator<Entry<Entity>> objectiterator = this.entitiesById.int2ObjectEntrySet().iterator();
 
-         label164:
-         while(true) {
-            Entity entity1;
-            while(true) {
-               if (!objectiterator.hasNext()) {
-                  this.field_217492_a = false;
 
-                  Entity entity;
-                  while((entity = this.field_217499_z.poll()) != null) {
-                     this.func_217465_m(entity);
-                  }
+         com.tuinity.tuinity.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator = this.entitiesForIteration.iterator(); // Tuinity
 
-                  this.func_217391_K();
-                  break label164;
-               }
+         while (objectiterator.hasNext()) {
+            Entity entity = (Entity) objectiterator.next(); // Tuinity
+            Entity entity1 = entity.func_184187_bx();
 
-               Entry<Entity> entry = objectiterator.next();
-               entity1 = entry.getValue();
-               Entity entity2 = entity1.func_184187_bx();
-               if (!this.field_73061_a.func_230537_U_() && (entity1 instanceof AnimalEntity || entity1 instanceof WaterMobEntity)) {
-                  entity1.func_70106_y();
-               }
+            iprofiler.func_76320_a("checkDespawn");
+            if (!entity.field_70128_L) {
+               entity.func_70623_bb();
+            }
+            iprofiler.func_76319_b();
 
-               if (!this.field_73061_a.func_230538_V_() && entity1 instanceof INPC) {
-                  entity1.func_70106_y();
+            if (entity1 != null) {
+               if (!entity1.field_70128_L && entity1.func_184196_w(entity)) {
+                  continue;
                }
 
-               iprofiler.func_76320_a("checkDespawn");
-               if (!entity1.field_70128_L) {
-                  entity1.func_70623_bb();
-               }
-
-               iprofiler.func_76319_b();
-               if (entity2 == null) {
-                  break;
-               }
-
-               if (entity2.field_70128_L || !entity2.func_184196_w(entity1)) {
-                  entity1.func_184210_p();
-                  break;
-               }
+               entity.func_184210_p();
             }
 
             iprofiler.func_76320_a("tick");
-            if (!entity1.field_70128_L && !(entity1 instanceof EnderDragonPartEntity)) {
-               this.func_217390_a(this::func_217479_a, entity1);
+            if (!entity.field_70128_L && !(entity instanceof EnderDragonPartEntity)) {
+               this.func_217390_a(this::func_217479_a, entity);
             }
-
             iprofiler.func_76319_b();
+
             iprofiler.func_76320_a("remove");
-            if (entity1.field_70128_L) {
-               this.func_217454_n(entity1);
-               objectiterator.remove();
-               this.func_217484_g(entity1);
+            if (entity.field_70128_L) {
+               this.func_217454_n(entity);
+               this.field_217498_x.remove(entity.func_145782_y());
+               this.removeEntityComplete(entity, entity instanceof ServerPlayerEntity); //Forge: Keep cap data until revive. Every other entity removes directly.
             }
-
             iprofiler.func_76319_b();
+
          }
+
+
+         objectiterator.finishedIterating(); // Tuinity
+         this.field_217492_a = false;
+         Entity entity2;
+
+         while ((entity2 = (Entity) this.field_217499_z.poll()) != null) {
+            if (!entity2.isQueuedForRegister) continue; // Paper - ignore cancelled registers
+            this.func_217465_m(entity2);
+         }
+
+         this.func_217391_K();
+         iprofiler.func_76319_b();
       }
 
       iprofiler.func_76319_b();
@@ -438,11 +435,23 @@
    }
 
    private void func_229856_ab_() {
-      this.field_217491_A.stream().filter(LivingEntity::func_70608_bn).collect(Collectors.toList()).forEach((p_241131_0_) -> {
-         p_241131_0_.func_225652_a_(false, false);
+      /*
+      this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach((p_241131_0_) -> {
+         p_241131_0_.stopSleepInBed(false, false);
       });
+       */
+      for (ServerPlayerEntity entityplayer : field_217491_A) { if (entityplayer.func_70608_bn()) { // Yatopia
+         entityplayer.func_225652_a_(false, false);
+      }} // Yatopia
    }
 
+   // Paper start - optimise random block ticking
+   private final BlockPos.Mutable chunkTickMutablePosition = new BlockPos.Mutable();
+   private final com.destroystokyo.paper.util.math.ThreadUnsafeRandom randomTickRandom = new com.destroystokyo.paper.util.math.ThreadUnsafeRandom();
+   // Paper end
+   private int currentIceAndSnowTick = 0; protected void resetIceAndSnowTick() { this.currentIceAndSnowTick = this.randomTickRandom.nextInt(16); } // AirplaneL
+
+
    public void func_217441_a(Chunk p_217441_1_, int p_217441_2_) {
       ChunkPos chunkpos = p_217441_1_.func_76632_l();
       boolean flag = this.func_72896_J();
@@ -450,69 +459,136 @@
       int j = chunkpos.func_180333_d();
       IProfiler iprofiler = this.func_217381_Z();
       iprofiler.func_76320_a("thunder");
-      if (flag && this.func_72911_I() && this.field_73012_v.nextInt(100000) == 0) {
-         BlockPos blockpos = this.func_175736_a(this.func_217383_a(i, 0, j, 15));
-         if (this.func_175727_C(blockpos)) {
-            DifficultyInstance difficultyinstance = this.func_175649_E(blockpos);
+      final BlockPos.Mutable blockposition = this.chunkTickMutablePosition; // Paper - use mutable to reduce allocation rate, final to force compile fail on change
+      if (flag && this.func_72911_I() && p_217441_1_.shouldDoLightning()) {
+         //BlockPos blockpos = this.adjustPosToNearbyEntity(this.getBlockRandomPos(i, 0, j, 15));
+         blockposition.func_189533_g(this.func_175736_a(this.func_217383_a(i, 0, j, 15)));
+         if (this.func_175727_C(blockposition)) {
+            DifficultyInstance difficultyinstance = this.func_175649_E(blockposition);
             boolean flag1 = this.func_82736_K().func_223586_b(GameRules.field_223601_d) && this.field_73012_v.nextDouble() < (double)difficultyinstance.func_180168_b() * 0.01D;
             if (flag1) {
                SkeletonHorseEntity skeletonhorseentity = EntityType.field_200742_ah.func_200721_a(this);
                skeletonhorseentity.func_190691_p(true);
                skeletonhorseentity.func_70873_a(0);
-               skeletonhorseentity.func_70107_b((double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p());
+               skeletonhorseentity.func_70107_b((double)blockposition.func_177958_n(), (double)blockposition.func_177956_o(), (double)blockposition.func_177952_p());
                this.func_217376_c(skeletonhorseentity);
             }
 
             LightningBoltEntity lightningboltentity = EntityType.field_200728_aG.func_200721_a(this);
-            lightningboltentity.func_233576_c_(Vector3d.func_237492_c_(blockpos));
+            lightningboltentity.func_233576_c_(Vector3d.func_237492_c_(blockposition));
             lightningboltentity.func_233623_a_(flag1);
             this.func_217376_c(lightningboltentity);
          }
       }
 
       iprofiler.func_219895_b("iceandsnow");
-      if (this.field_73012_v.nextInt(16) == 0) {
-         BlockPos blockpos2 = this.func_205770_a(Heightmap.Type.MOTION_BLOCKING, this.func_217383_a(i, 0, j, 15));
-         BlockPos blockpos3 = blockpos2.func_177977_b();
-         Biome biome = this.func_226691_t_(blockpos2);
-         if (biome.func_201848_a(this, blockpos3)) {
-            this.func_175656_a(blockpos3, Blocks.field_150432_aD.func_176223_P());
+      if (/* this.rand.nextInt(16) == 0*/ (this.currentIceAndSnowTick++ & 15) == 0) {
+         /*
+         BlockPos blockpos2 = this.getHeight(Heightmap.Type.MOTION_BLOCKING, this.getBlockRandomPos(i, 0, j, 15));
+         BlockPos blockpos3 = blockpos2.down();
+
+          */
+
+         // Paper start - optimise chunk ticking
+         this.getRandomBlockPosition(i, 0, j, 15, blockposition);
+         int normalY = p_217441_1_.func_201576_a(Heightmap.Type.MOTION_BLOCKING, blockposition.func_177958_n() & 15, blockposition.func_177952_p() & 15);
+         int downY = normalY - 1;
+         blockposition.func_185336_p(normalY);
+         // Paper end
+
+         Biome biome = this.func_226691_t_(blockposition);
+
+         blockposition.func_185336_p(downY);
+         if (this.isAreaLoaded(blockposition, 1)) // Forge: check area to avoid loading neighbors in unloaded chunks
+         if (biome.func_201848_a(this, blockposition)) {
+            this.func_175656_a(blockposition, Blocks.field_150432_aD.func_176223_P());
          }
 
-         if (flag && biome.func_201850_b(this, blockpos2)) {
-            this.func_175656_a(blockpos2, Blocks.field_150433_aE.func_176223_P());
+         blockposition.func_185336_p(downY);
+         if (flag && biome.func_201850_b(this, blockposition)) {
+            this.func_175656_a(blockposition, Blocks.field_150433_aE.func_176223_P());
          }
 
-         if (flag && this.func_226691_t_(blockpos3).func_201851_b() == Biome.RainType.RAIN) {
-            this.func_180495_p(blockpos3).func_177230_c().func_176224_k(this, blockpos3);
+         blockposition.func_185336_p(downY);
+         if (flag && this.func_226691_t_(blockposition).func_201851_b() == Biome.RainType.RAIN) {
+            this.func_180495_p(blockposition).func_177230_c().func_176224_k(this, blockposition);
          }
       }
 
       iprofiler.func_219895_b("tickBlocks");
-      if (p_217441_2_ > 0) {
-         for(ChunkSection chunksection : p_217441_1_.func_76587_i()) {
-            if (chunksection != Chunk.field_186036_a && chunksection.func_206915_b()) {
-               int k = chunksection.func_222632_g();
 
-               for(int l = 0; l < p_217441_2_; ++l) {
-                  BlockPos blockpos1 = this.func_217383_a(i, k, j, 15);
-                  iprofiler.func_76320_a("randomTick");
-                  BlockState blockstate = chunksection.func_177485_a(blockpos1.func_177958_n() - i, blockpos1.func_177956_o() - k, blockpos1.func_177952_p() - j);
-                  if (blockstate.func_204519_t()) {
-                     blockstate.func_227034_b_(this, blockpos1, this.field_73012_v);
+      ChunkSection[] sections = p_217441_1_.func_76587_i();
+
+      for (int sectionIndex = 0; sectionIndex < 16; ++sectionIndex) {
+         ChunkSection section = sections[sectionIndex];
+         if (section == null || section.tickingList.size() == 0 || !section.func_76675_b()) {
+            continue;
+         }
+
+         int yPos = sectionIndex << 4;
+
+         for (int a = 0; a < p_217441_2_; ++a) {
+            iprofiler.func_76320_a("randomTick");
+            int tickingBlocks = section.tickingList.size();
+            int index = this.randomTickRandom.nextInt(16 * 16 * 16);
+            if (index >= tickingBlocks) {
+               continue;
+            }
+
+            long raw = section.tickingList.getRaw(index);
+            int location = com.destroystokyo.paper.util.maplist.IBlockDataList.getLocationFromRaw(raw);
+            int randomX = location & 15;
+            int randomY = ((location >>> (4 + 4)) & 255) | yPos;
+            int randomZ = (location >>> 4) & 15;
+
+            BlockState iblockdata = com.destroystokyo.paper.util.maplist.IBlockDataList.getBlockDataFromRaw(raw);
+            BlockPos blockposition2 = blockposition.func_181079_c(i + randomX, randomY, j + randomZ);
+
+            if (iblockdata.func_204519_t()) {
+               iblockdata.func_227034_b_(this, blockposition2, this.randomTickRandom);
+            }
+
+            FluidState fluidstate = iblockdata.func_204520_s();
+            if (fluidstate.func_206890_h()) {
+               fluidstate.func_206891_b(this, blockposition2, this.randomTickRandom);
+            }
+            /*
+
+            iblockdata.tick(this, blockposition2, this.randomTickRandom);
+
+             */
+
+            // We drop the fluid tick since LAVA is ALREADY TICKED by the above method.
+            // TODO CHECK ON UPDATE
+            iprofiler.func_76319_b();
+         }
+      }
+      /*
+      if (randomTickSpeed > 0) {
+         for(ChunkSection chunksection : chunkIn.getSections()) {
+            if (chunksection != Chunk.EMPTY_SECTION && chunksection.needsRandomTickAny()) {
+               int k = chunksection.getYLocation();
+
+               for(int l = 0; l < randomTickSpeed; ++l) {
+                  BlockPos blockpos1 = this.getBlockRandomPos(i, k, j, 15);
+                  iprofiler.startSection("randomTick");
+                  BlockState blockstate = chunksection.getBlockState(blockpos1.getX() - i, blockpos1.getY() - k, blockpos1.getZ() - j);
+                  if (blockstate.ticksRandomly()) {
+                     blockstate.randomTick(this, blockpos1, this.rand);
                   }
 
-                  FluidState fluidstate = blockstate.func_204520_s();
-                  if (fluidstate.func_206890_h()) {
-                     fluidstate.func_206891_b(this, blockpos1, this.field_73012_v);
+                  FluidState fluidstate = blockstate.getFluidState();
+                  if (fluidstate.ticksRandomly()) {
+                     fluidstate.randomTick(this, blockpos1, this.rand);
                   }
 
-                  iprofiler.func_76319_b();
+                  iprofiler.endSection();
                }
             }
          }
       }
 
+       */
       iprofiler.func_76319_b();
    }
 
@@ -598,9 +674,10 @@
             ++p_217479_1_.field_70173_aa;
             IProfiler iprofiler = this.func_217381_Z();
             iprofiler.func_194340_a(() -> {
-               return Registry.field_212629_r.func_177774_c(p_217479_1_.func_200600_R()).toString();
+               return p_217479_1_.func_200600_R().getRegistryName() == null ? p_217479_1_.func_200600_R().toString() : p_217479_1_.func_200600_R().getRegistryName().toString();
             });
             iprofiler.func_230035_c_("tickNonPassenger");
+            if (p_217479_1_.canUpdate())
             p_217479_1_.func_70071_h_();
             iprofiler.func_76319_b();
          }
@@ -675,6 +752,38 @@
       return !this.field_73061_a.func_175579_a(this, p_175660_2_, p_175660_1_) && this.func_175723_af().func_177746_a(p_175660_2_);
    }
 
+   // Paper start - derived from below
+   public void saveIncrementally(boolean doFull) {
+      ServerChunkProvider chunkproviderserver = this.func_72863_F();
+
+      if (doFull) {
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Save(this));
+      }
+
+      try  {
+         if (doFull) {
+            this.func_73042_a();
+         }
+
+         if (!this.func_217402_u()) chunkproviderserver.saveIncrementally();
+
+
+         // Copied from save()
+         // CraftBukkit start - moved from MinecraftServer.saveChunks
+         if (doFull) { // Paper
+            ServerWorld worldserver1 = this;
+
+            field_241103_E_.func_230393_a_(worldserver1.func_175723_af().func_235927_t_());
+            field_73061_a.field_240768_i_.func_230414_b_(this.field_73061_a.func_201300_aS().func_201380_c());
+            field_73061_a.getAnvilConverterForAnvilFile().func_237288_a_(this.field_73061_a.getDynRegistry(), this.field_73061_a.field_240768_i_, this.field_73061_a.func_184103_al().func_72378_q());
+         }
+         // CraftBukkit end
+      } catch (Exception e) {
+         e.printStackTrace();
+      }
+   }
+   // Paper end
+
    public void func_217445_a(@Nullable IProgressUpdate p_217445_1_, boolean p_217445_2_, boolean p_217445_3_) {
       ServerChunkProvider serverchunkprovider = this.func_72863_F();
       if (!p_217445_3_) {
@@ -687,6 +796,7 @@
             p_217445_1_.func_200209_c(new TranslationTextComponent("menu.savingChunks"));
          }
 
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Save(this));
          serverchunkprovider.func_217210_a(p_217445_2_);
       }
    }
@@ -777,6 +887,7 @@
    }
 
    private void func_217448_f(ServerPlayerEntity p_217448_1_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(p_217448_1_, this))) return;
       Entity entity = this.field_175741_N.get(p_217448_1_.func_110124_au());
       if (entity != null) {
          field_147491_a.warn("Force-added player with duplicate UUID {}", (Object)p_217448_1_.func_110124_au().toString());
@@ -801,6 +912,7 @@
       } else if (this.func_217478_l(p_72838_1_)) {
          return false;
       } else {
+         if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(p_72838_1_, this))) return false;
          IChunk ichunk = this.func_217353_a(MathHelper.func_76128_c(p_72838_1_.func_226277_ct_() / 16.0D), MathHelper.func_76128_c(p_72838_1_.func_226281_cx_() / 16.0D), ChunkStatus.field_222617_m, p_72838_1_.field_98038_p);
          if (!(ichunk instanceof Chunk)) {
             return false;
@@ -816,6 +928,7 @@
       if (this.func_217478_l(p_217440_1_)) {
          return false;
       } else {
+         if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(p_217440_1_, this))) return false;
          this.func_217465_m(p_217440_1_);
          return true;
       }
@@ -879,12 +992,17 @@
 
    }
 
+   @Deprecated //Forge: Use removeEntityComplete(entity,boolean)
    public void func_217484_g(Entity p_217484_1_) {
+      removeEntityComplete(p_217484_1_, false);
+   }
+   public void removeEntityComplete(Entity p_217484_1_, boolean keepData) {
       if (p_217484_1_ instanceof EnderDragonEntity) {
          for(EnderDragonPartEntity enderdragonpartentity : ((EnderDragonEntity)p_217484_1_).func_213404_dT()) {
-            enderdragonpartentity.func_70106_y();
+            enderdragonpartentity.remove(keepData);
          }
       }
+      p_217484_1_.remove(keepData);
 
       this.field_175741_N.remove(p_217484_1_.func_110124_au());
       this.func_72863_F().func_217226_b(p_217484_1_);
@@ -898,35 +1016,47 @@
          this.field_217495_I.remove(((MobEntity)p_217484_1_).func_70661_as());
       }
 
+      p_217484_1_.onRemovedFromWorld();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityLeaveWorldEvent(p_217484_1_, this));
    }
 
    private void func_217465_m(Entity p_217465_1_) {
       if (this.field_217492_a) {
          this.field_217499_z.add(p_217465_1_);
+         p_217465_1_.isQueuedForRegister = true; // Paper
       } else {
+         p_217465_1_.isQueuedForRegister = false; // Paper
          this.field_217498_x.put(p_217465_1_.func_145782_y(), p_217465_1_);
+         this.entitiesForIteration.add(p_217465_1_); // Tuinity
          if (p_217465_1_ instanceof EnderDragonEntity) {
             for(EnderDragonPartEntity enderdragonpartentity : ((EnderDragonEntity)p_217465_1_).func_213404_dT()) {
                this.field_217498_x.put(enderdragonpartentity.func_145782_y(), enderdragonpartentity);
+               this.entitiesForIteration.add(enderdragonpartentity); // Tuinity
             }
          }
 
          this.field_175741_N.put(p_217465_1_.func_110124_au(), p_217465_1_);
+         this.entitiesForIteration.add(p_217465_1_); // Tuinity
          this.func_72863_F().func_217230_c(p_217465_1_);
          if (p_217465_1_ instanceof MobEntity) {
             this.field_217495_I.add(((MobEntity)p_217465_1_).func_70661_as());
          }
       }
 
+      p_217465_1_.onAddedToWorld();
    }
 
    public void func_217467_h(Entity p_217467_1_) {
-      if (this.field_217492_a) {
+      removeEntity(p_217467_1_, false);
+   }
+   public void removeEntity(Entity p_217467_1_, boolean keepData) {
+      if (false && this.field_217492_a) {
          throw (IllegalStateException)Util.func_229757_c_(new IllegalStateException("Removing entity while ticking!"));
       } else {
          this.func_217454_n(p_217467_1_);
          this.field_217498_x.remove(p_217467_1_.func_145782_y());
-         this.func_217484_g(p_217467_1_);
+         this.entitiesForIteration.remove(p_217467_1_); // Tuinity
+         this.removeEntityComplete(p_217467_1_, keepData);
       }
    }
 
@@ -939,8 +1069,11 @@
    }
 
    public void func_217434_e(ServerPlayerEntity p_217434_1_) {
-      p_217434_1_.func_70106_y();
-      this.func_217467_h(p_217434_1_);
+      removePlayer(p_217434_1_, false);
+   }
+   public void removePlayer(ServerPlayerEntity p_217434_1_, boolean keepData) {
+      p_217434_1_.remove(keepData);
+      this.removeEntity(p_217434_1_, keepData);
       this.func_72854_c();
    }
 
@@ -959,10 +1092,20 @@
    }
 
    public void func_184148_a(@Nullable PlayerEntity p_184148_1_, double p_184148_2_, double p_184148_4_, double p_184148_6_, SoundEvent p_184148_8_, SoundCategory p_184148_9_, float p_184148_10_, float p_184148_11_) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(p_184148_1_, p_184148_8_, p_184148_9_, p_184148_10_, p_184148_11_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_184148_8_ = event.getSound();
+      p_184148_9_ = event.getCategory();
+      p_184148_10_ = event.getVolume();
       this.field_73061_a.func_184103_al().func_148543_a(p_184148_1_, p_184148_2_, p_184148_4_, p_184148_6_, p_184148_10_ > 1.0F ? (double)(16.0F * p_184148_10_) : 16.0D, this.func_234923_W_(), new SPlaySoundEffectPacket(p_184148_8_, p_184148_9_, p_184148_2_, p_184148_4_, p_184148_6_, p_184148_10_, p_184148_11_));
    }
 
    public void func_217384_a(@Nullable PlayerEntity p_217384_1_, Entity p_217384_2_, SoundEvent p_217384_3_, SoundCategory p_217384_4_, float p_217384_5_, float p_217384_6_) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(p_217384_1_, p_217384_3_, p_217384_4_, p_217384_5_, p_217384_6_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_217384_3_ = event.getSound();
+      p_217384_4_ = event.getCategory();
+      p_217384_5_ = event.getVolume();
       this.field_73061_a.func_184103_al().func_148543_a(p_217384_1_, p_217384_2_.func_226277_ct_(), p_217384_2_.func_226278_cu_(), p_217384_2_.func_226281_cx_(), p_217384_5_ > 1.0F ? (double)(16.0F * p_217384_5_) : 16.0D, this.func_234923_W_(), new SSpawnMovingSoundEffectPacket(p_217384_3_, p_217384_4_, p_217384_2_, p_217384_5_, p_217384_6_));
    }
 
@@ -998,6 +1141,7 @@
 
    public Explosion func_230546_a_(@Nullable Entity p_230546_1_, @Nullable DamageSource p_230546_2_, @Nullable ExplosionContext p_230546_3_, double p_230546_4_, double p_230546_6_, double p_230546_8_, float p_230546_10_, boolean p_230546_11_, Explosion.Mode p_230546_12_) {
       Explosion explosion = new Explosion(this, p_230546_1_, p_230546_2_, p_230546_3_, p_230546_4_, p_230546_6_, p_230546_8_, p_230546_10_, p_230546_11_, p_230546_12_);
+      if (net.minecraftforge.event.ForgeEventFactory.onExplosionStart(this, explosion)) return explosion;
       explosion.func_77278_a();
       explosion.func_77279_a(false);
       if (p_230546_12_ == Explosion.Mode.NONE) {
@@ -1410,4 +1554,14 @@
          p_241121_0_.func_175656_a(p_241122_1_, Blocks.field_150343_Z.func_176223_P());
       });
    }
+
+   protected void initCapabilities() {
+      this.gatherCapabilities();
+      capabilityData = this.func_217481_x().func_215752_a(() -> new net.minecraftforge.common.util.WorldCapabilityData(getCapabilities()), net.minecraftforge.common.util.WorldCapabilityData.ID);
+      capabilityData.setCapabilities(getCapabilities());
+   }
+
+   public java.util.stream.Stream<Entity> getEntities() {
+       return field_217498_x.values().stream();
+   }
 }
