--- a/net/minecraft/world/server/ServerChunkProvider.java
+++ b/net/minecraft/world/server/ServerChunkProvider.java
@@ -131,7 +131,7 @@
             return p_222874_0_;
          }, (p_222870_1_) -> {
             if (p_212849_4_) {
-               throw (IllegalStateException)Util.func_229757_c_(new IllegalStateException("Chunk not there when requested: " + p_222870_1_));
+               throw Util.func_229757_c_(new IllegalStateException("Chunk not there when requested: " + p_222870_1_));
             } else {
                return null;
             }
@@ -346,14 +346,14 @@
          EntityClassification[] aentityclassification = EntityClassification.values();
          Object2IntMap<EntityClassification> object2intmap = this.field_73251_h.func_217450_l();
          this.field_73251_h.func_217381_Z().func_76319_b();
-         this.field_217237_a.func_223491_f().forEach((p_223434_10_) -> {
-            Optional<Chunk> optional = p_223434_10_.func_219297_b().getNow(ChunkHolder.field_219308_c).left();
+         this.field_217237_a.func_223491_f().forEach((chunkHolder) -> {
+            Optional<Chunk> optional = chunkHolder.func_219297_b().getNow(ChunkHolder.field_219308_c).left();
             if (optional.isPresent()) {
                Chunk chunk = optional.get();
                this.field_73251_h.func_217381_Z().func_76320_a("broadcast");
-               p_223434_10_.func_219274_a(chunk);
+               chunkHolder.func_219274_a(chunk);
                this.field_73251_h.func_217381_Z().func_76319_b();
-               ChunkPos chunkpos = p_223434_10_.func_219277_h();
+               ChunkPos chunkpos = chunkHolder.func_219277_h();
\ No newline at end of file
                if (!this.field_217237_a.func_219243_d(chunkpos)) {
                   chunk.func_177415_c(chunk.func_177416_w() + j);
                   if (flag1 && (this.field_217246_l || this.field_217247_m) && this.field_73251_h.func_175723_af().func_177730_a(chunk.func_76632_l())) {
