--- a/net/minecraft/world/server/ServerTickList.java
+++ b/net/minecraft/world/server/ServerTickList.java
@@ -134,6 +134,7 @@
    private List<NextTickListEntry<T>> func_223187_a(@Nullable List<NextTickListEntry<T>> p_223187_1_, Collection<NextTickListEntry<T>> p_223187_2_, MutableBoundingBox p_223187_3_, boolean p_223187_4_) {
       Iterator<NextTickListEntry<T>> iterator = p_223187_2_.iterator();
 
+
       while(iterator.hasNext()) {
          NextTickListEntry<T> nextticklistentry = iterator.next();
          BlockPos blockpos = nextticklistentry.field_180282_a;
