--- a/net/minecraft/util/concurrent/ITaskQueue.java
+++ b/net/minecraft/util/concurrent/ITaskQueue.java
@@ -25,8 +25,7 @@
          }).collect(Collectors.toList());
       }
 
-      @Nullable
-      public Runnable func_212827_a_() {
+         public Runnable func_212827_a_() {
          for(Queue<Runnable> queue : this.field_219949_a) {
             Runnable runnable = queue.poll();
             if (runnable != null) {
@@ -73,8 +72,7 @@
          this.field_219947_a = p_i50962_1_;
       }
 
-      @Nullable
-      public T func_212827_a_() {
+         public T func_212827_a_() {
          return this.field_219947_a.poll();
       }
 
