--- a/net/minecraft/item/crafting/RecipeManager.java
+++ b/net/minecraft/item/crafting/RecipeManager.java
@@ -18,6 +18,9 @@
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+
+import fr.modcraftmc.forge.SilentException;
+import fr.modcraftmc.forge.config.ModcraftConfig;
 import net.minecraft.client.resources.JsonReloadListener;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.item.ItemStack;
@@ -50,9 +53,18 @@
 
       for(Entry<ResourceLocation, JsonObject> entry : p_212853_1_.entrySet()) {
          ResourceLocation resourcelocation = entry.getKey();
+         if (resourcelocation.func_110623_a().startsWith("_")) continue; //Forge: filter anything beginning with "_" as it's used for metadata.
 
          try {
+            if (!net.minecraftforge.common.crafting.CraftingHelper.processConditions(entry.getValue(), "conditions")) {
+               field_199521_c.debug("Skipping loading recipe {} as it's conditions were not met", resourcelocation);
+               continue;
+            }
             IRecipe<?> irecipe = func_215377_a(resourcelocation, entry.getValue());
+            if (irecipe == null) {
+               if (ModcraftConfig.isShowRecipeManagerError()) field_199521_c.info("Skipping loading recipe {} as it's serializer returned null", resourcelocation);
+               continue;
+            }
             map.computeIfAbsent(irecipe.func_222127_g(), (p_223391_0_) -> {
                return ImmutableMap.builder();
             }).put(resourcelocation, irecipe);
@@ -121,7 +133,8 @@
    public static IRecipe<?> func_215377_a(ResourceLocation p_215377_0_, JsonObject p_215377_1_) {
       String s = JSONUtils.func_151200_h(p_215377_1_, "type");
       return Registry.field_218368_I.func_218349_b(new ResourceLocation(s)).orElseThrow(() -> {
-         return new JsonSyntaxException("Invalid or unsupported recipe type '" + s + "'");
+         if (ModcraftConfig.isShowRecipeManagerError()) return new JsonSyntaxException("Invalid or unsupported recipe type '" + s + "'");
+         else return new SilentException();
       }).func_199425_a_(p_215377_0_, p_215377_1_);
    }
 
