--- a/net/minecraft/util/AxisRotation.java
+++ b/net/minecraft/util/AxisRotation.java
@@ -20,7 +20,20 @@
       }
 
       public Direction.Axis func_197513_a(Direction.Axis p_197513_1_) {
-         return field_197521_d[Math.floorMod(p_197513_1_.ordinal() + 1, 3)];
+        // return AXES[Math.floorMod(axisIn.ordinal() + 1, 3)];
+
+         switch (p_197513_1_) {
+            case X:
+               return Direction.Axis.Y;
+            case Y:
+               return Direction.Axis.Z;
+            case Z:
+               return Direction.Axis.X;
+         }
+
+         throw new IllegalArgumentException();
+         // Yatopia end
+
       }
 
       public AxisRotation func_197514_a() {
@@ -33,7 +46,19 @@
       }
 
       public Direction.Axis func_197513_a(Direction.Axis p_197513_1_) {
-         return field_197521_d[Math.floorMod(p_197513_1_.ordinal() - 1, 3)];
+         //return AXES[Math.floorMod(axisIn.ordinal() - 1, 3)];
+
+         switch (p_197513_1_) {
+            case X:
+               return Direction.Axis.Z;
+            case Y:
+               return Direction.Axis.X;
+            case Z:
+               return Direction.Axis.Y;
+         }
+
+         throw new IllegalArgumentException();
+         // Yatopia end
       }
 
       public AxisRotation func_197514_a() {
