--- a/net/minecraft/util/SharedConstants.java
+++ b/net/minecraft/util/SharedConstants.java
@@ -10,6 +10,8 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
+import java.util.function.IntPredicate;
+
 public class SharedConstants {
    public static final Level field_184877_a = Level.DISABLED;
    public static boolean field_206244_b;
@@ -20,6 +22,11 @@
       return p_71566_0_ != 167 && p_71566_0_ >= ' ' && p_71566_0_ != 127;
    }
 
+   public static boolean isDisAllowedCharacterInString(String string) {
+
+      return (string.chars().anyMatch(value -> value == 167) || string.chars().anyMatch(value -> value == 127));
+   }
+
    public static String func_71565_a(String p_71565_0_) {
       StringBuilder stringbuilder = new StringBuilder();
 
@@ -57,6 +64,7 @@
    }
 
    static {
+      if (System.getProperty("io.netty.leakDetection.level") == null) // Forge: allow level to be manually specified
       ResourceLeakDetector.setLevel(field_184877_a);
       CommandSyntaxException.ENABLE_COMMAND_STACK_TRACES = false;
       CommandSyntaxException.BUILT_IN_EXCEPTIONS = new TranslatableExceptionProvider();
