--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -174,7 +174,7 @@
       throw p_229756_0_ instanceof RuntimeException ? (RuntimeException)p_229756_0_ : new RuntimeException(p_229756_0_);
    }
 
-   private static void func_240983_a_(Thread p_240983_0_, Throwable p_240983_1_) {
+   public static void func_240983_a_(Thread p_240983_0_, Throwable p_240983_1_) {
       func_229757_c_(p_240983_1_);
       if (p_240983_1_ instanceof CompletionException) {
          p_240983_1_ = p_240983_1_.getCause();
