--- a/net/minecraft/network/NettyCompressionDecoder.java
+++ b/net/minecraft/network/NettyCompressionDecoder.java
@@ -28,7 +28,8 @@
                throw new DecoderException("Badly compressed packet - size of " + i + " is below server threshold of " + this.field_179304_b);
             }
 
-            if (i > 2097152) {
+            //2097152
+            if (i > 16777216) {
\ No newline at end of file
                throw new DecoderException("Badly compressed packet - size of " + i + " is larger than protocol maximum of " + 2097152);
             }
 
