--- a/net/minecraft/command/impl/StopCommand.java
+++ b/net/minecraft/command/impl/StopCommand.java
@@ -3,7 +3,11 @@
 import com.mojang.brigadier.CommandDispatcher;
 import net.minecraft.command.CommandSource;
 import net.minecraft.command.Commands;
+import net.minecraft.entity.player.ServerPlayerEntity;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraftforge.fml.server.ServerLifecycleHooks;
+import org.lwjgl.system.CallbackI;
 
 public class StopCommand {
    public static void func_198725_a(CommandDispatcher<CommandSource> p_198725_0_) {
