--- a/net/minecraft/command/impl/SummonCommand.java
+++ b/net/minecraft/command/impl/SummonCommand.java
@@ -47,7 +47,7 @@
          return 1;
       } else {
          ServerWorld serverworld = p_198737_0_.func_197023_e();
-         Entity entity = EntityType.func_220335_a(compoundnbt, serverworld, (p_218914_2_) -> {
+         Entity entity = EntityType.getPassenger(compoundnbt, serverworld, (p_218914_2_) -> {
             p_218914_2_.func_70012_b(p_198737_2_.field_72450_a, p_198737_2_.field_72448_b, p_198737_2_.field_72449_c, p_218914_2_.field_70177_z, p_218914_2_.field_70125_A);
             return !serverworld.func_217470_d(p_218914_2_) ? null : p_218914_2_;
          });
