--- a/net/minecraft/util/palette/PaletteArray.java
+++ b/net/minecraft/util/palette/PaletteArray.java
@@ -1,6 +1,7 @@
 package net.minecraft.util.palette;
 
 import java.util.function.Function;
+import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.ListNBT;
@@ -47,6 +48,17 @@
       return ArrayUtils.contains(this.field_186042_a, p_222626_1_);
    }
 
+   @Override
+   public boolean accepts(Predicate<T> predicate) {
+      for(int i = 0; i < this.field_186045_d; ++i) {
+         if (predicate.test(this.field_186042_a[i])) {
+            return true;
+         }
+      }
+
+      return false;
+   }
+
    @Nullable
    public T func_186039_a(int p_186039_1_) {
       return (T)(p_186039_1_ >= 0 && p_186039_1_ < this.field_186045_d ? this.field_186042_a[p_186039_1_] : null);
