--- a/net/minecraft/world/IEntityReader.java
+++ b/net/minecraft/world/IEntityReader.java
@@ -6,17 +6,20 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityPredicate;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.util.EntityPredicates;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.shapes.IBooleanFunction;
 import net.minecraft.util.math.shapes.VoxelShape;
 import net.minecraft.util.math.shapes.VoxelShapes;
+import net.minecraft.world.server.ServerWorld;
 
 public interface IEntityReader {
    List<Entity> func_175674_a(@Nullable Entity p_175674_1_, AxisAlignedBB p_175674_2_, @Nullable Predicate<? super Entity> p_175674_3_);
@@ -179,6 +182,24 @@
       return list;
    }
 
+   default Stream<Entity> getEntitiesInRange(Entity player, ServerWorld world, int range) {
+
+      return world.getEntities()
+              .filter(entity -> !entity.equals(player))
+              .filter(entity -> player.func_70032_d(entity) <= range);
+
+     // return list;
+   }
+
+   default Stream<Entity> getParallelEntitiesInRange(Entity player, ServerWorld world, int range) {
+
+      return world.getEntities().parallel()
+              .filter(entity -> !entity.equals(player))
+              .filter(entity -> player.func_70032_d(entity) <= range);
+
+      // return list;
+   }
+
\ No newline at end of file
    default <T extends LivingEntity> List<T> func_217374_a(Class<? extends T> p_217374_1_, EntityPredicate p_217374_2_, LivingEntity p_217374_3_, AxisAlignedBB p_217374_4_) {
       List<T> list = this.func_175647_a(p_217374_1_, p_217374_4_, (Predicate<T>)null);
       List<T> list1 = Lists.newArrayList();
