--- a/net/minecraft/world/chunk/ChunkPrimerWrapper.java
+++ b/net/minecraft/world/chunk/ChunkPrimerWrapper.java
@@ -1,5 +1,7 @@
 package net.minecraft.world.chunk;
 
+import ca.starlight.common.chunk.ExtendedChunk;
+import ca.starlight.common.light.SWMRNibbleArray;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import java.util.BitSet;
 import java.util.Map;
@@ -22,7 +24,7 @@
 import net.minecraft.world.gen.feature.structure.StructureStart;
 import net.minecraft.world.lighting.WorldLightManager;
 
-public class ChunkPrimerWrapper extends ChunkPrimer {
+public class ChunkPrimerWrapper extends ChunkPrimer implements IChunk, ExtendedChunk {
    private final Chunk field_209220_a;
 
    public ChunkPrimerWrapper(Chunk p_i49948_1_) {
@@ -30,6 +32,31 @@
       this.field_209220_a = p_i49948_1_;
    }
 
+   @Override
+   public SWMRNibbleArray[] getBlockNibbles() {
+      return ((ExtendedChunk)this.field_209220_a).getBlockNibbles();
+   }
+
+   @Override
+   public void setBlockNibbles(final SWMRNibbleArray[] nibbles) {
+      ((ExtendedChunk)this.field_209220_a).setBlockNibbles(nibbles);
+   }
+
+   @Override
+   public SWMRNibbleArray[] getSkyNibbles() {
+      return ((ExtendedChunk)this.field_209220_a).getSkyNibbles();
+   }
+
+   @Override
+   public void setSkyNibbles(final SWMRNibbleArray[] nibbles) {
+      ((ExtendedChunk)this.field_209220_a).setSkyNibbles(nibbles);
+   }
+
+   @Override
+   public boolean[][] getEmptinessMap() {
+      return ((ExtendedChunk)this.field_209220_a).getEmptinessMap();
+   }
+
    @Nullable
    public TileEntity func_175625_s(BlockPos p_175625_1_) {
       return this.field_209220_a.func_175625_s(p_175625_1_);
