--- a/net/minecraft/network/play/ServerPlayNetHandler.java
+++ b/net/minecraft/network/play/ServerPlayNetHandler.java
@@ -4,10 +4,14 @@
 import com.google.common.primitives.Floats;
 import com.mojang.brigadier.ParseResults;
 import com.mojang.brigadier.StringReader;
+import fr.modcraftmc.modcraftforge.theading.ModcraftThreadFactory;
+import fr.modcraftmc.modcraftforge.utils.Utils;
 import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.GenericFutureListener;
 import it.unimi.dsi.fastutil.ints.Int2ShortMap;
 import it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap;
+
+import java.text.DecimalFormat;
 import java.util.Collections;
 import java.util.Set;
 import javax.annotation.Nullable;
@@ -31,11 +35,7 @@
 import net.minecraft.entity.player.PlayerInventory;
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.AbstractArrowEntity;
-import net.minecraft.inventory.container.BeaconContainer;
-import net.minecraft.inventory.container.Container;
-import net.minecraft.inventory.container.MerchantContainer;
-import net.minecraft.inventory.container.RecipeBookContainer;
-import net.minecraft.inventory.container.RepairContainer;
+import net.minecraft.inventory.container.*;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
 import net.minecraft.item.WritableBookItem;
@@ -89,17 +89,7 @@
 import net.minecraft.network.play.client.CUpdateSignPacket;
 import net.minecraft.network.play.client.CUpdateStructureBlockPacket;
 import net.minecraft.network.play.client.CUseEntityPacket;
-import net.minecraft.network.play.server.SChangeBlockPacket;
-import net.minecraft.network.play.server.SChatPacket;
-import net.minecraft.network.play.server.SConfirmTransactionPacket;
-import net.minecraft.network.play.server.SDisconnectPacket;
-import net.minecraft.network.play.server.SHeldItemChangePacket;
-import net.minecraft.network.play.server.SKeepAlivePacket;
-import net.minecraft.network.play.server.SMoveVehiclePacket;
-import net.minecraft.network.play.server.SPlayerPositionLookPacket;
-import net.minecraft.network.play.server.SQueryNBTResponsePacket;
-import net.minecraft.network.play.server.SSetSlotPacket;
-import net.minecraft.network.play.server.STabCompletePacket;
+import net.minecraft.network.play.server.*;
 import net.minecraft.potion.Effects;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.CommandBlockLogic;
@@ -129,6 +119,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.dimension.DimensionType;
 import net.minecraft.world.server.ServerWorld;
+import fr.modcraftmc.modcraftforge.ModcraftConfig;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -167,6 +158,13 @@
    private int field_184347_F;
    private int field_184348_G;
 
+   //ModcraftForge
+   private int tickCount = 0;
+   private StringTextComponent header;
+   private StringTextComponent footer;
+
+   private static final DecimalFormat TIME_FORMATTER = new DecimalFormat("########0.000");
+
    public ServerPlayNetHandler(MinecraftServer p_i1530_1_, NetworkManager p_i1530_2_, ServerPlayerEntity p_i1530_3_) {
       this.field_147367_d = p_i1530_1_;
       this.field_147371_a = p_i1530_2_;
@@ -221,6 +219,7 @@
 
       this.field_147367_d.func_213185_aS().func_76320_a("keepAlive");
       long i = Util.func_211177_b();
+
       if (i - this.field_194402_f >= 15000L) {
          if (this.field_194403_g) {
             this.func_194028_b(new TranslationTextComponent("disconnect.timeout"));
@@ -245,8 +244,38 @@
          this.func_194028_b(new TranslationTextComponent("multiplayer.disconnect.idling"));
       }
 
+      if (!ModcraftConfig.enableCustomTablist()) return;
+
+      if (++tickCount > ModcraftConfig.getRefreshRate()) {
+         tickCount = 0;
+         this.processTabUpdate();
+      }
+
    }
 
+   public void processTabUpdate() {
+
+      double meanTickTime = Utils.mean(this.field_147367_d.field_71311_j) * 1.0E-6D;
+      double meanTPS = Math.min(1000.0/meanTickTime, 20);
+
+      header = new StringTextComponent(ModcraftConfig.getCustomheader()
+              .replaceAll("%TPS%", TIME_FORMATTER.format(meanTPS))
+              .replaceAll("%TICKTIME%", TIME_FORMATTER.format(meanTickTime) + "MS")
+              .replaceAll("%ONLINE%", String.valueOf(field_147367_d.func_184103_al().func_72394_k()))
+              .replaceAll("%MAX%", String.valueOf(field_147367_d.func_184103_al().func_72352_l()))
+              .replaceAll("%PING%", String.valueOf(field_147369_b.field_71138_i)));
+
+
+      footer = new StringTextComponent(ModcraftConfig.getCustomFooter()
+              .replaceAll("%TPS%", TIME_FORMATTER.format(meanTPS))
+              .replaceAll("%TICKTIME%", TIME_FORMATTER.format(meanTickTime) + "MS")
+              .replaceAll("%ONLINE%", String.valueOf(field_147367_d.func_184103_al().func_72394_k()))
+              .replaceAll("%MAX%", String.valueOf(field_147367_d.func_184103_al().func_72352_l()))
+              .replaceAll("%PING%", String.valueOf(field_147369_b.field_71138_i)));
+
+      field_147369_b.field_71135_a.func_147359_a(new SPlayerListHeaderFooterPacket(header, footer));
+   }
+
    public void func_184342_d() {
       this.field_184349_l = this.field_147369_b.func_226277_ct_();
       this.field_184350_m = this.field_147369_b.func_226278_cu_();
@@ -336,9 +365,11 @@
             }
 
             entity.func_70080_a(d3, d4, d5, f, f1);
+            this.field_147369_b.func_70080_a(d3, d4, d5, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
             boolean flag2 = serverworld.func_226665_a__(entity, entity.func_174813_aQ().func_186664_h(0.0625D));
             if (flag && (flag1 || !flag2)) {
                entity.func_70080_a(d0, d1, d2, f, f1);
+               this.field_147369_b.func_70080_a(d3, d4, d5, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
                this.field_147371_a.func_179290_a(new SMoveVehiclePacket(entity));
                return;
             }
@@ -713,11 +744,17 @@
 
                      if (!this.field_147369_b.func_184850_K() && (!this.field_147369_b.func_71121_q().func_82736_K().func_223586_b(GameRules.field_223615_r) || !this.field_147369_b.func_184613_cA())) {
                         float f2 = this.field_147369_b.func_184613_cA() ? 300.0F : 100.0F;
+
+                        /*ModcraftForge start - juste disable it
+
                         if (d11 - d10 > (double)(f2 * (float)i) && !this.func_217264_d()) {
-                           field_147370_c.warn("{} moved too quickly! {},{},{}", this.field_147369_b.func_200200_C_().getString(), d7, d8, d9);
-                           this.func_147364_a(this.field_147369_b.func_226277_ct_(), this.field_147369_b.func_226278_cu_(), this.field_147369_b.func_226281_cx_(), this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
+                           LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getName().getString(), d7, d8, d9);
+                           this.setPlayerLocation(this.player.getPosX(), this.player.getPosY(), this.player.getPosZ(), this.player.rotationYaw, this.player.rotationPitch);
                            return;
                         }
+
+                       ModcraftForge end
+                         */
                      }
 
                      boolean flag2 = this.func_223133_a(serverworld);
@@ -845,7 +882,9 @@
       Direction direction = blockraytraceresult.func_216354_b();
       this.field_147369_b.func_143004_u();
       if (blockpos.func_177956_o() < this.field_147367_d.func_71207_Z() - 1 || direction != Direction.UP && blockpos.func_177956_o() < this.field_147367_d.func_71207_Z()) {
-         if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < 64.0D && serverworld.func_175660_a(this.field_147369_b, blockpos)) {
+         double dist = field_147369_b.func_110148_a(net.minecraft.entity.player.PlayerEntity.REACH_DISTANCE).func_111126_e() + 3;
+         dist *= dist;
+         if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < dist && serverworld.func_175660_a(this.field_147369_b, blockpos)) {
             ActionResultType actionresulttype = this.field_147369_b.field_71134_c.func_219441_a(this.field_147369_b, serverworld, itemstack, hand, blockraytraceresult);
             if (actionresulttype.func_226247_b_()) {
                this.field_147369_b.func_226292_a_(hand, true);
@@ -900,7 +939,19 @@
    public void func_147231_a(ITextComponent p_147231_1_) {
       field_147370_c.info("{} lost connection: {}", this.field_147369_b.func_200200_C_().getString(), p_147231_1_.getString());
       this.field_147367_d.func_147132_au();
-      this.field_147367_d.func_184103_al().func_148539_a((new TranslationTextComponent("multiplayer.player.left", this.field_147369_b.func_145748_c_())).func_211708_a(TextFormatting.YELLOW));
+
+      ITextComponent component;
+
+      if (ModcraftConfig.enableCustomConnexionMessages()) {
+         component = new StringTextComponent(String.format(ModcraftConfig.getCustomLogoutMessage(), this.field_147369_b.func_145748_c_().func_150254_d()));
+
+      } else {
+         component = new TranslationTextComponent("multiplayer.player.left", this.field_147369_b.func_145748_c_()).func_211708_a(TextFormatting.YELLOW);
+      }
+
+
+      this.field_147367_d.func_184103_al().func_148539_a(component);
+
       this.field_147369_b.func_71123_m();
       this.field_147367_d.func_184103_al().func_72367_e(this.field_147369_b);
       if (this.func_217264_d()) {
@@ -911,7 +962,7 @@
    }
 
    public void func_147359_a(IPacket<?> p_147359_1_) {
-      this.func_211148_a(p_147359_1_, (GenericFutureListener<? extends Future<? super Void>>)null);
+      this.func_211148_a(p_147359_1_, null);
    }
 
    public void func_211148_a(IPacket<?> p_211148_1_, @Nullable GenericFutureListener<? extends Future<? super Void>> p_211148_2_) {
@@ -945,39 +996,50 @@
          this.field_147369_b.field_71071_by.field_70461_c = p_147355_1_.func_149614_c();
          this.field_147369_b.func_143004_u();
       } else {
-         field_147370_c.warn("{} tried to set an invalid carried item", (Object)this.field_147369_b.func_200200_C_().getString());
+         field_147370_c.warn("{} tried to set an invalid carried item", this.field_147369_b.func_200200_C_().getString());
       }
    }
 
    public void func_147354_a(CChatMessagePacket p_147354_1_) {
-      PacketThreadUtil.func_218796_a(p_147354_1_, this, this.field_147369_b.func_71121_q());
-      if (this.field_147369_b.func_147096_v() == ChatVisibility.HIDDEN) {
-         this.func_147359_a(new SChatPacket((new TranslationTextComponent("chat.cannotSend")).func_211708_a(TextFormatting.RED)));
-      } else {
-         this.field_147369_b.func_143004_u();
-         String s = p_147354_1_.func_149439_c();
-         s = org.apache.commons.lang3.StringUtils.normalizeSpace(s);
+      //PacketThreadUtil.checkThreadAndEnqueue(packetIn, this, this.player.getServerWorld());
 
-         for(int i = 0; i < s.length(); ++i) {
-            if (!SharedConstants.func_71566_a(s.charAt(i))) {
-               this.func_194028_b(new TranslationTextComponent("multiplayer.disconnect.illegal_characters"));
-               return;
-            }
-         }
+      ModcraftThreadFactory.CHAT_THREAD.submit(() -> {
 
-         if (s.startsWith("/")) {
-            this.func_147361_d(s);
+         if (this.field_147369_b.func_147096_v() == ChatVisibility.HIDDEN) {
+            this.func_147359_a(new SChatPacket((new TranslationTextComponent("chat.cannotSend")).func_211708_a(TextFormatting.RED)));
          } else {
-            ITextComponent itextcomponent = new TranslationTextComponent("chat.type.text", this.field_147369_b.func_145748_c_(), s);
-            this.field_147367_d.func_184103_al().func_148544_a(itextcomponent, false);
-         }
 
-         this.field_147374_l += 20;
-         if (this.field_147374_l > 200 && !this.field_147367_d.func_184103_al().func_152596_g(this.field_147369_b.func_146103_bH())) {
-            this.func_194028_b(new TranslationTextComponent("disconnect.spam"));
+            this.field_147369_b.func_143004_u();
+            String message = p_147354_1_.func_149439_c();
+            message = org.apache.commons.lang3.StringUtils.normalizeSpace(message);
+
+            for (int i = 0; i < message.length(); ++i) {
+               if (!SharedConstants.func_71566_a(message.charAt(i))) {
+                  this.func_194028_b(new TranslationTextComponent("multiplayer.disconnect.illegal_characters"));
+                  return;
+               }
+            }
+
+
+            if (message.startsWith("/")) {
+               String finalMessage = message;
+               //run command on server thread
+               field_147367_d.func_213167_f(()-> this.func_147361_d(finalMessage));
+            } else {
+               ITextComponent itextcomponent = new TranslationTextComponent("chat.type.text", this.field_147369_b.func_145748_c_(), net.minecraftforge.common.ForgeHooks.newChatWithLinks(message));
+               itextcomponent = net.minecraftforge.common.ForgeHooks.onServerChatEvent(this, message, itextcomponent);
+               if (itextcomponent == null) return;
+               this.field_147367_d.func_184103_al().func_148544_a(itextcomponent, false);
+            }
+
+            this.field_147374_l += 20;
+            if (this.field_147374_l > 200 && !this.field_147367_d.func_184103_al().func_152596_g(this.field_147369_b.func_146103_bH())) {
+               this.func_194028_b(new TranslationTextComponent("disconnect.spam"));
+            }
+
          }
 
-      }
+      });
    }
 
    private void func_147361_d(String p_147361_1_) {
@@ -1061,6 +1123,7 @@
                this.field_147369_b.func_190775_a(entity, hand);
             } else if (p_147340_1_.func_149565_c() == CUseEntityPacket.Action.INTERACT_AT) {
                Hand hand1 = p_147340_1_.func_186994_b();
+               if (net.minecraftforge.common.ForgeHooks.onInteractEntityAt(field_147369_b, entity, p_147340_1_.func_179712_b(), hand1) != null) return;
                ActionResultType actionresulttype = entity.func_184199_a(this.field_147369_b, p_147340_1_.func_179712_b(), hand1);
                if (actionresulttype.func_226247_b_()) {
                   this.field_147369_b.func_226292_a_(hand1, true);
@@ -1094,7 +1157,7 @@
                return;
             }
 
-            this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, DimensionType.field_223227_a_, false);
+            this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, this.field_147369_b.field_71093_bK, false);
             if (this.field_147367_d.func_71199_h()) {
                this.field_147369_b.func_71033_a(GameType.SPECTATOR);
                this.field_147369_b.func_71121_q().func_82736_K().func_223585_a(GameRules.field_223613_p).func_223570_a(false, this.field_147367_d);
@@ -1119,10 +1182,17 @@
          if (this.field_147369_b.func_175149_v()) {
             NonNullList<ItemStack> nonnulllist = NonNullList.func_191196_a();
 
-            for(int i = 0; i < this.field_147369_b.field_71070_bA.field_75151_b.size(); ++i) {
-               nonnulllist.add(this.field_147369_b.field_71070_bA.field_75151_b.get(i).func_75211_c());
+
+            /*
+            for(int i = 0; i < this.player.openContainer.inventorySlots.size(); ++i) {
+               nonnulllist.add(this.player.openContainer.inventorySlots.get(i).getStack());
             }
+             */
 
+            for (Slot slot : this.field_147369_b.field_71070_bA.field_75151_b) {
+               nonnulllist.add(slot.func_75211_c());
+            }
+
             this.field_147369_b.func_71110_a(this.field_147369_b.field_71070_bA, nonnulllist);
          } else {
             ItemStack itemstack1 = this.field_147369_b.field_71070_bA.func_184996_a(p_147351_1_.func_149544_d(), p_147351_1_.func_149543_e(), p_147351_1_.func_186993_f(), this.field_147369_b);
@@ -1138,11 +1208,13 @@
                this.field_147369_b.field_71070_bA.func_75128_a(this.field_147369_b, false);
                NonNullList<ItemStack> nonnulllist1 = NonNullList.func_191196_a();
 
-               for(int j = 0; j < this.field_147369_b.field_71070_bA.field_75151_b.size(); ++j) {
+
+                for(int j = 0; j < this.field_147369_b.field_71070_bA.field_75151_b.size(); ++j) {
                   ItemStack itemstack = this.field_147369_b.field_71070_bA.field_75151_b.get(j).func_75211_c();
                   nonnulllist1.add(itemstack.func_190926_b() ? ItemStack.field_190927_a : itemstack);
                }
 
+
                this.field_147369_b.func_71110_a(this.field_147369_b.field_71070_bA, nonnulllist1);
             }
          }
@@ -1268,6 +1340,8 @@
    }
 
    public void func_147349_a(CCustomPayloadPacket p_147349_1_) {
+      PacketThreadUtil.func_218796_a(p_147349_1_, this, this.field_147369_b.func_71121_q());
+      net.minecraftforge.fml.network.NetworkHooks.onCustomPayload(p_147349_1_, this.field_147371_a);
    }
 
    public void func_217263_a(CSetDifficultyPacket p_217263_1_) {
@@ -1283,4 +1357,5 @@
          this.field_147367_d.func_213209_d(p_217261_1_.func_218776_b());
       }
    }
+
 }
