--- a/net/minecraft/client/multiplayer/ClientChunkProvider.java
+++ b/net/minecraft/client/multiplayer/ClientChunkProvider.java
@@ -59,6 +59,7 @@
          int i = this.field_217256_d.func_217191_a(p_73234_1_, p_73234_2_);
          Chunk chunk = this.field_217256_d.func_217192_a(i);
          if (func_217249_a(chunk, p_73234_1_, p_73234_2_)) {
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload(chunk));
             this.field_217256_d.func_217190_a(i, chunk, (Chunk)null);
          }
 
@@ -112,6 +113,7 @@
          }
 
          this.field_73235_d.func_228323_e_(p_228313_1_, p_228313_2_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
          return chunk;
       }
    }
@@ -220,8 +222,7 @@
          return Math.abs(p_217183_1_ - this.field_217198_e) <= this.field_217196_c && Math.abs(p_217183_2_ - this.field_217199_f) <= this.field_217196_c;
       }
 
-      @Nullable
-      protected Chunk func_217192_a(int p_217192_1_) {
+         protected Chunk func_217192_a(int p_217192_1_) {
          return this.field_217195_b.get(p_217192_1_);
       }
    }
