--- a/net/minecraft/advancements/AdvancementRewards.java
+++ b/net/minecraft/advancements/AdvancementRewards.java
@@ -39,7 +39,7 @@
 
    public void func_192113_a(ServerPlayerEntity p_192113_1_) {
       p_192113_1_.func_195068_e(this.field_192115_b);
-      LootContext lootcontext = (new LootContext.Builder(p_192113_1_.func_71121_q())).func_216015_a(LootParameters.field_216281_a, p_192113_1_).func_216015_a(LootParameters.field_237457_g_, p_192113_1_.func_213303_ch()).func_216023_a(p_192113_1_.func_70681_au()).func_216022_a(LootParameterSets.field_216265_f);
+      LootContext lootcontext = (new LootContext.Builder(p_192113_1_.func_71121_q())).func_216015_a(LootParameters.field_216281_a, p_192113_1_).func_216015_a(LootParameters.field_237457_g_, p_192113_1_.func_213303_ch()).func_216023_a(p_192113_1_.func_70681_au()).func_186469_a(p_192113_1_.func_184817_da()).func_216022_a(LootParameterSets.field_216265_f); // FORGE: luck to LootContext
       boolean flag = false;
 
       for(ResourceLocation resourcelocation : this.field_192116_c) {
@@ -142,8 +142,7 @@
       private int field_200282_a;
       private final List<ResourceLocation> field_200283_b = Lists.newArrayList();
       private final List<ResourceLocation> field_200284_c = Lists.newArrayList();
-      @Nullable
-      private ResourceLocation field_200285_d;
+         private ResourceLocation field_200285_d;
 
       public static AdvancementRewards.Builder func_203907_a(int p_203907_0_) {
          return (new AdvancementRewards.Builder()).func_203906_b(p_203907_0_);
