--- a/net/minecraft/server/management/PlayerList.java
+++ b/net/minecraft/server/management/PlayerList.java
@@ -7,6 +7,7 @@
 import io.netty.buffer.Unpooled;
 import java.io.File;
 import java.net.SocketAddress;
+import java.text.DecimalFormat;
 import java.text.SimpleDateFormat;
 import java.util.List;
 import java.util.Map;
@@ -24,25 +25,7 @@
 import net.minecraft.network.NetworkManager;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.ServerPlayNetHandler;
-import net.minecraft.network.play.server.SChangeGameStatePacket;
-import net.minecraft.network.play.server.SChatPacket;
-import net.minecraft.network.play.server.SCustomPayloadPlayPacket;
-import net.minecraft.network.play.server.SEntityStatusPacket;
-import net.minecraft.network.play.server.SHeldItemChangePacket;
-import net.minecraft.network.play.server.SJoinGamePacket;
-import net.minecraft.network.play.server.SPlayEntityEffectPacket;
-import net.minecraft.network.play.server.SPlayerAbilitiesPacket;
-import net.minecraft.network.play.server.SPlayerListItemPacket;
-import net.minecraft.network.play.server.SRespawnPacket;
-import net.minecraft.network.play.server.SServerDifficultyPacket;
-import net.minecraft.network.play.server.SSetExperiencePacket;
-import net.minecraft.network.play.server.SSpawnPositionPacket;
-import net.minecraft.network.play.server.STagsListPacket;
-import net.minecraft.network.play.server.STeamsPacket;
-import net.minecraft.network.play.server.SUpdateRecipesPacket;
-import net.minecraft.network.play.server.SUpdateTimePacket;
-import net.minecraft.network.play.server.SUpdateViewDistancePacket;
-import net.minecraft.network.play.server.SWorldBorderPacket;
+import net.minecraft.network.play.server.*;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.scoreboard.ScoreObjective;
 import net.minecraft.scoreboard.ScorePlayerTeam;
@@ -53,10 +36,7 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
-import net.minecraft.util.text.ChatType;
-import net.minecraft.util.text.ITextComponent;
-import net.minecraft.util.text.TextFormatting;
-import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraft.util.text.*;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.GameType;
 import net.minecraft.world.IWorld;
@@ -94,7 +74,11 @@
    private GameType field_72410_m;
    private boolean field_72407_n;
    private int field_72408_o;
+   private final List<ServerPlayerEntity> playersView = java.util.Collections.unmodifiableList(field_72404_b);
 
+   private static final DecimalFormat TIME_FORMATTER = new DecimalFormat("########0.000");
+   private static final long[] UNLOADED = new long[] {0};
+
    public PlayerList(MinecraftServer p_i50688_1_, int p_i50688_2_) {
       this.field_72400_f = p_i50688_1_;
       this.field_72405_c = p_i50688_2_;
@@ -109,7 +93,15 @@
       String s = gameprofile1 == null ? gameprofile.getName() : gameprofile1.getName();
       playerprofilecache.func_152649_a(gameprofile);
       CompoundNBT compoundnbt = this.func_72380_a(p_72355_2_);
-      ServerWorld serverworld = this.field_72400_f.func_71218_a(p_72355_2_.field_71093_bK);
+
+      //Forge: Make sure the dimension hasn't been deleted, if so stick them in the overworld.
+      ServerWorld serverworld = p_72355_2_.field_71093_bK != null ? this.field_72400_f.func_71218_a(p_72355_2_.field_71093_bK) : null ;
+      if (serverworld == null) {
+         p_72355_2_.field_71093_bK = DimensionType.field_223227_a_;
+         serverworld = this.field_72400_f.func_71218_a(p_72355_2_.field_71093_bK);
+         p_72355_2_.func_70107_b(serverworld.func_72912_H().func_76079_c(), serverworld.func_72912_H().func_76075_d(), serverworld.func_72912_H().func_76074_e());
+      }
+
       p_72355_2_.func_70029_a(serverworld);
       p_72355_2_.field_71134_c.func_73080_a((ServerWorld)p_72355_2_.field_70170_p);
       String s1 = "local";
@@ -121,6 +113,8 @@
       WorldInfo worldinfo = serverworld.func_72912_H();
       this.func_72381_a(p_72355_2_, (ServerPlayerEntity)null, serverworld);
       ServerPlayNetHandler serverplaynethandler = new ServerPlayNetHandler(this.field_72400_f, p_72355_1_, p_72355_2_);
+      net.minecraftforge.fml.network.NetworkHooks.sendMCRegistryPackets(p_72355_1_, "PLAY_TO_CLIENT");
+      net.minecraftforge.fml.network.NetworkHooks.sendDimensionDataPacket(p_72355_1_, p_72355_2_);
       GameRules gamerules = serverworld.func_82736_K();
       boolean flag = gamerules.func_223586_b(GameRules.field_226683_z_);
       boolean flag1 = gamerules.func_223586_b(GameRules.field_223612_o);
@@ -131,6 +125,13 @@
       serverplaynethandler.func_147359_a(new SHeldItemChangePacket(p_72355_2_.field_71071_by.field_70461_c));
       serverplaynethandler.func_147359_a(new SUpdateRecipesPacket(this.field_72400_f.func_199529_aN().func_199510_b()));
       serverplaynethandler.func_147359_a(new STagsListPacket(this.field_72400_f.func_199731_aO()));
+
+      serverplaynethandler.func_147359_a(new SPlayerListHeaderFooterPacket(new StringTextComponent("§aBienvenue sur https://modcraftmc.fr \n §7Skyblock moddé 1.15.2 ! \n §8https://discord.me/modcraftmc"),
+              new StringTextComponent("§8TPS: loading..." + "\n Ticktime: loading...")));
+
+      //         this.sendPacketToAllPlayers(new SPlayerListHeaderFooterPacket(new StringTextComponent("§aBienvenue sur https://modcraftmc.fr \n §7Skyblock moddé 1.15.2 ! \n §8https://discord.me/modcraftmc"),
+      //                 new StringTextComponent("§8TPS: "+ meanTPS + "\n Ticktime: " + meanTickTime)));
+
       this.func_187243_f(p_72355_2_);
       p_72355_2_.func_147099_x().func_150877_d();
       p_72355_2_.func_192037_E().func_192826_c(p_72355_2_);
@@ -145,12 +146,12 @@
 
       this.func_148539_a(itextcomponent.func_211708_a(TextFormatting.YELLOW));
       serverplaynethandler.func_147364_a(p_72355_2_.func_226277_ct_(), p_72355_2_.func_226278_cu_(), p_72355_2_.func_226281_cx_(), p_72355_2_.field_70177_z, p_72355_2_.field_70125_A);
-      this.field_72404_b.add(p_72355_2_);
+      this.addPlayer(p_72355_2_);
       this.field_177454_f.put(p_72355_2_.func_110124_au(), p_72355_2_);
       this.func_148540_a(new SPlayerListItemPacket(SPlayerListItemPacket.Action.ADD_PLAYER, p_72355_2_));
 
-      for(int i = 0; i < this.field_72404_b.size(); ++i) {
-         p_72355_2_.field_71135_a.func_147359_a(new SPlayerListItemPacket(SPlayerListItemPacket.Action.ADD_PLAYER, this.field_72404_b.get(i)));
+      for (ServerPlayerEntity player : this.field_72404_b) {
+         p_72355_2_.field_71135_a.func_147359_a(new SPlayerListItemPacket(SPlayerListItemPacket.Action.ADD_PLAYER, player));
       }
 
       serverworld.func_217435_c(p_72355_2_);
@@ -166,8 +167,9 @@
 
       if (compoundnbt != null && compoundnbt.func_150297_b("RootVehicle", 10)) {
          CompoundNBT compoundnbt1 = compoundnbt.func_74775_l("RootVehicle");
+         final ServerWorld worldf = serverworld;
          Entity entity1 = EntityType.func_220335_a(compoundnbt1.func_74775_l("Entity"), serverworld, (p_217885_1_) -> {
-            return !serverworld.func_217470_d(p_217885_1_) ? null : p_217885_1_;
+            return !worldf.func_217470_d(p_217885_1_) ? null : p_217885_1_;
          });
          if (entity1 != null) {
             UUID uuid = compoundnbt1.func_186857_a("Attach");
@@ -194,6 +196,7 @@
       }
 
       p_72355_2_.func_71116_b();
+      net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerLoggedIn( p_72355_2_ );
    }
 
    protected void func_96456_a(ServerScoreboard p_96456_1_, ServerPlayerEntity p_96456_2_) {
@@ -255,6 +258,7 @@
          compoundnbt1 = compoundnbt;
          p_72380_1_.func_70020_e(compoundnbt);
          field_148546_d.debug("loading single player");
+         net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(p_72380_1_, this.field_72412_k, p_72380_1_.func_110124_au().toString());
       } else {
          compoundnbt1 = this.field_72412_k.func_75752_b(p_72380_1_);
       }
@@ -263,6 +267,7 @@
    }
 
    protected void func_72391_b(ServerPlayerEntity p_72391_1_) {
+      if (p_72391_1_.field_71135_a == null) return;
       this.field_72412_k.func_75753_a(p_72391_1_);
       ServerStatisticsManager serverstatisticsmanager = this.field_148547_k.get(p_72391_1_.func_110124_au());
       if (serverstatisticsmanager != null) {
@@ -277,6 +282,7 @@
    }
 
    public void func_72367_e(ServerPlayerEntity p_72367_1_) {
+      net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerLoggedOut(p_72367_1_);
       ServerWorld serverworld = p_72367_1_.func_71121_q();
       p_72367_1_.func_195066_a(Stats.field_75947_j);
       this.func_72391_b(p_72367_1_);
@@ -298,7 +304,7 @@
       p_72367_1_.func_213319_R();
       serverworld.func_217434_e(p_72367_1_);
       p_72367_1_.func_192039_O().func_192745_a();
-      this.field_72404_b.remove(p_72367_1_);
+      this.removePlayer(p_72367_1_);
       this.field_72400_f.func_201300_aS().func_201382_b(p_72367_1_);
       UUID uuid = p_72367_1_.func_110124_au();
       ServerPlayerEntity serverplayerentity = this.field_177454_f.get(uuid);
@@ -367,10 +373,18 @@
    }
 
    public ServerPlayerEntity func_72368_a(ServerPlayerEntity p_72368_1_, DimensionType p_72368_2_, boolean p_72368_3_) {
-      this.field_72404_b.remove(p_72368_1_);
-      p_72368_1_.func_71121_q().func_217434_e(p_72368_1_);
-      BlockPos blockpos = p_72368_1_.func_180470_cg();
-      boolean flag = p_72368_1_.func_82245_bX();
+      ServerWorld world = field_72400_f.func_71218_a(p_72368_2_);
+      if (world == null)
+         p_72368_2_ = p_72368_1_.getSpawnDimension();
+      else if (!world.func_201675_m().func_76567_e())
+         p_72368_2_ = world.func_201675_m().getRespawnDimension(p_72368_1_);
+      if (field_72400_f.func_71218_a(p_72368_2_) == null)
+         p_72368_2_ = DimensionType.field_223227_a_;
+
+      this.removePlayer(p_72368_1_);
+      p_72368_1_.func_71121_q().removePlayer(p_72368_1_, true); // Forge: keep data until copyFrom called
+      BlockPos blockpos = p_72368_1_.getBedLocation(p_72368_2_);
+      boolean flag = p_72368_1_.isSpawnForced(p_72368_2_);
       p_72368_1_.field_71093_bK = p_72368_2_;
       PlayerInteractionManager playerinteractionmanager;
       if (this.field_72400_f.func_71242_L()) {
@@ -382,6 +396,8 @@
       ServerPlayerEntity serverplayerentity = new ServerPlayerEntity(this.field_72400_f, this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK), p_72368_1_.func_146103_bH(), playerinteractionmanager);
       serverplayerentity.field_71135_a = p_72368_1_.field_71135_a;
       serverplayerentity.func_193104_a(p_72368_1_, p_72368_3_);
+      p_72368_1_.remove(false); // Forge: clone event had a chance to see old data, now discard it
+      serverplayerentity.field_71093_bK = p_72368_2_;
       serverplayerentity.func_145769_d(p_72368_1_.func_145782_y());
       serverplayerentity.func_184819_a(p_72368_1_.func_184591_cq());
 
@@ -396,7 +412,7 @@
          if (optional.isPresent()) {
             Vec3d vec3d = optional.get();
             serverplayerentity.func_70012_b(vec3d.field_72450_a, vec3d.field_72448_b, vec3d.field_72449_c, 0.0F, 0.0F);
-            serverplayerentity.func_226560_a_(blockpos, flag, false);
+            serverplayerentity.setSpawnPoint(blockpos, flag, false, p_72368_2_);
          } else {
             serverplayerentity.field_71135_a.func_147359_a(new SChangeGameStatePacket(0, 0.0F));
          }
@@ -407,6 +423,7 @@
       }
 
       WorldInfo worldinfo = serverplayerentity.field_70170_p.func_72912_H();
+      net.minecraftforge.fml.network.NetworkHooks.sendDimensionDataPacket(serverplayerentity.field_71135_a.field_147371_a, serverplayerentity);
       serverplayerentity.field_71135_a.func_147359_a(new SRespawnPacket(serverplayerentity.field_71093_bK, WorldInfo.func_227498_c_(worldinfo.func_76063_b()), worldinfo.func_76067_t(), serverplayerentity.field_71134_c.func_73081_b()));
       BlockPos blockpos1 = serverworld.func_175694_M();
       serverplayerentity.field_71135_a.func_147364_a(serverplayerentity.func_226277_ct_(), serverplayerentity.func_226278_cu_(), serverplayerentity.func_226281_cx_(), serverplayerentity.field_70177_z, serverplayerentity.field_70125_A);
@@ -416,10 +433,11 @@
       this.func_72354_b(serverplayerentity, serverworld);
       this.func_187243_f(serverplayerentity);
       serverworld.func_217433_d(serverplayerentity);
-      this.field_72404_b.add(serverplayerentity);
+      this.addPlayer(serverplayerentity);
       this.field_177454_f.put(serverplayerentity.func_110124_au(), serverplayerentity);
       serverplayerentity.func_71116_b();
       serverplayerentity.func_70606_j(serverplayerentity.func_110143_aJ());
+      net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerRespawnEvent(serverplayerentity, p_72368_3_);
       return serverplayerentity;
    }
 
@@ -432,6 +450,13 @@
    public void func_72374_b() {
       if (++this.field_72408_o > 600) {
          this.func_148540_a(new SPlayerListItemPacket(SPlayerListItemPacket.Action.UPDATE_LATENCY, this.field_72404_b));
+
+         double meanTickTime = mean(field_72400_f.field_71311_j) * 1.0E-6D;
+         double meanTPS = Math.min(1000.0/meanTickTime, 20);
+
+         this.func_148540_a(new SPlayerListHeaderFooterPacket(new StringTextComponent("§aBienvenue sur https://modcraftmc.fr \n §7Skyblock moddé 1.15.2 ! \n §8https://discord.me/modcraftmc"),
+                 new StringTextComponent("§8TPS: "+ meanTPS + "\n Ticktime: " + meanTickTime)));
+
          this.field_72408_o = 0;
       }
 
@@ -739,7 +764,7 @@
    }
 
    public List<ServerPlayerEntity> func_181057_v() {
-      return this.field_72404_b;
+      return this.playersView; //Unmodifiable view, we don't want people removing things without us knowing.
    }
 
    @Nullable
@@ -769,4 +794,20 @@
    public boolean func_206257_x() {
       return this.field_72407_n;
    }
+
+   public boolean addPlayer(ServerPlayerEntity player) {
+      return net.minecraftforge.common.DimensionManager.rebuildPlayerMap(this, this.field_72404_b.add(player));
+   }
+
+   public boolean removePlayer(ServerPlayerEntity player) {
+       return net.minecraftforge.common.DimensionManager.rebuildPlayerMap(this, this.field_72404_b.remove(player));
+   }
+
+   private static long mean(long[] values)
+   {
+      long sum = 0L;
+      for (long v : values)
+         sum += v;
+      return sum / values.length;
+   }
 }
