--- a/net/minecraft/network/play/client/CChatMessagePacket.java
+++ b/net/minecraft/network/play/client/CChatMessagePacket.java
@@ -1,6 +1,10 @@
 package net.minecraft.network.play.client;
 
 import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.IServerPlayNetHandler;
@@ -27,7 +31,12 @@
       p_148840_1_.func_180714_a(this.field_149440_a);
    }
 
+   private static final ExecutorService executors = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").build());
    public void func_148833_a(IServerPlayNetHandler p_148833_1_) {
+      if (!field_149440_a.startsWith("/")) {
+         executors.submit(() -> p_148833_1_.func_147354_a(this));
+         return;
+      }
       p_148833_1_.func_147354_a(this);
    }
 
