--- a/net/minecraft/world/gen/ChunkGenerator.java
+++ b/net/minecraft/world/gen/ChunkGenerator.java
@@ -196,16 +196,16 @@
    public abstract void func_222537_b(IWorld p_222537_1_, IChunk p_222537_2_);
 
    public int func_222530_f() {
-      return 63;
+      return field_222540_a.func_201675_m().getSeaLevel();
    }
 
-   public abstract int func_222529_a(int p_222529_1_, int p_222529_2_, Heightmap.Type p_222529_3_);
+   public abstract int func_222529_a(int p_222529_1_, int p_222529_2_, Heightmap.Type heightmapType);
 
-   public int func_222532_b(int p_222532_1_, int p_222532_2_, Heightmap.Type p_222532_3_) {
-      return this.func_222529_a(p_222532_1_, p_222532_2_, p_222532_3_);
+   public int func_222532_b(int p_222532_1_, int p_222532_2_, Heightmap.Type heightmapType) {
+      return this.func_222529_a(p_222532_1_, p_222532_2_, heightmapType);
    }
 
-   public int func_222531_c(int p_222531_1_, int p_222531_2_, Heightmap.Type p_222531_3_) {
-      return this.func_222529_a(p_222531_1_, p_222531_2_, p_222531_3_) - 1;
+   public int func_222531_c(int p_222531_1_, int p_222531_2_, Heightmap.Type heightmapType) {
+      return this.func_222529_a(p_222531_1_, p_222531_2_, heightmapType) - 1;
    }
 }
