--- a/net/minecraft/world/gen/NoiseChunkGenerator.java
+++ b/net/minecraft/world/gen/NoiseChunkGenerator.java
@@ -140,7 +140,7 @@
       return 0.0D;
    }
 
-   public int func_222529_a(int p_222529_1_, int p_222529_2_, Heightmap.Type p_222529_3_) {
+   public int func_222529_a(int p_222529_1_, int p_222529_2_, Heightmap.Type heightmapType) {
       int i = Math.floorDiv(p_222529_1_, this.field_222564_k);
       int j = Math.floorDiv(p_222529_2_, this.field_222564_k);
       int k = Math.floorMod(p_222529_1_, this.field_222564_k);
@@ -172,7 +172,7 @@
                   blockstate = this.field_222560_g;
                }
 
-               if (p_222529_3_.func_222684_d().test(blockstate)) {
+               if (heightmapType.func_222684_d().test(blockstate)) {
                   return l1 + 1;
                }
             }
@@ -378,8 +378,8 @@
                         }
 
                         if (blockstate != field_222562_i) {
-                           if (blockstate.func_185906_d() != 0) {
-                              blockpos$mutable.func_181079_c(j3, j2, i4);
+                           blockpos$mutable.func_181079_c(j3, j2, i4);
+                           if (blockstate.getLightValue(chunkprimer, blockpos$mutable) != 0) {
                               chunkprimer.func_201637_h(blockpos$mutable);
                            }
 
