--- a/net/minecraft/server/ServerEula.java
+++ b/net/minecraft/server/ServerEula.java
@@ -27,7 +27,7 @@
          return flag;
       } catch (Exception var16) {
          field_154349_a.warn("Failed to load {}", (Object)this.field_154350_b);
-         this.func_154348_b();
+         this.createEULAFile(false);
          return false;
       }
    }
@@ -36,11 +36,11 @@
       return this.field_154351_c;
    }
 
-   private void func_154348_b() {
+   public void createEULAFile(boolean state) {
       if (!SharedConstants.field_206244_b) {
          try (OutputStream outputstream = Files.newOutputStream(this.field_154350_b)) {
             Properties properties = new Properties();
-            properties.setProperty("eula", "false");
+            properties.setProperty("eula", String.valueOf(state));
\ No newline at end of file
             properties.store(outputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
          } catch (Exception exception) {
             field_154349_a.warn("Failed to save {}", this.field_154350_b, exception);
