--- a/net/minecraft/util/math/AxisAlignedBB.java
+++ b/net/minecraft/util/math/AxisAlignedBB.java
@@ -39,12 +39,39 @@
       return new AxisAlignedBB((double)p_216363_0_.field_78897_a, (double)p_216363_0_.field_78895_b, (double)p_216363_0_.field_78896_c, (double)(p_216363_0_.field_78893_d + 1), (double)(p_216363_0_.field_78894_e + 1), (double)(p_216363_0_.field_78892_f + 1));
    }
 
+   /**
+    * @reason Simplify the code to better help the JVM optimize it
+    * @author JellySquid
+    */
    public double func_197745_a(Direction.Axis p_197745_1_) {
-      return p_197745_1_.func_196051_a(this.field_72340_a, this.field_72338_b, this.field_72339_c);
+      /*return axis.getCoordinate(this.minX, this.minY, this.minZ);*/
+      switch (p_197745_1_) {
+         case X:
+            return this.field_72340_a;
+         case Y:
+            return this.field_72338_b;
+         case Z:
+            return this.field_72339_c;
+      }
+      throw new IllegalArgumentException();
    }
 
+   /**
+    * @reason Simplify the code to better help the JVM optimize it
+    * @author JellySquid
+    */
    public double func_197742_b(Direction.Axis p_197742_1_) {
-      return p_197742_1_.func_196051_a(this.field_72336_d, this.field_72337_e, this.field_72334_f);
+      /*return axis.getCoordinate(this.maxX, this.maxY, this.maxZ);*/
+      switch (p_197742_1_) {
+         case X:
+            return this.field_72336_d;
+         case Y:
+            return this.field_72337_e;
+         case Z:
+            return this.field_72334_f;
+      }
+
+      throw new IllegalArgumentException();
    }
 
    public boolean equals(Object p_equals_1_) {
