--- a/net/minecraft/data/SingleItemRecipeBuilder.java
+++ b/net/minecraft/data/SingleItemRecipeBuilder.java
@@ -103,13 +103,11 @@
          return this.field_218627_h;
       }
 
-      @Nullable
-      public JsonObject func_200440_c() {
+         public JsonObject func_200440_c() {
          return this.field_218625_f.func_200273_b();
       }
 
-      @Nullable
-      public ResourceLocation func_200443_d() {
+         public ResourceLocation func_200443_d() {
          return this.field_218626_g;
       }
    }
